sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/AccessTokenCacheImpl.java

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.containers.containerregistry.implementation.authentication;
+
+import com.azure.core.credential.AccessToken;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Signal;
+import reactor.core.publisher.Sinks;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * A token cache that supports caching a token and refreshing it.
+ */
+class AccessTokenCacheImpl {
+    // The delay after a refresh to attempt another token refresh
+    private static final Duration REFRESH_DELAY = Duration.ofSeconds(30);
+    // the offset before token expiry to attempt proactive token refresh
+    private static final Duration REFRESH_OFFSET = Duration.ofMinutes(5);
+    private volatile AccessToken cache;
+    private volatile OffsetDateTime nextTokenRefresh = OffsetDateTime.now();
+    private final AtomicReference<Sinks.One<AccessToken>> wip;
+    private final ContainerRegistryTokenCredentialImpl tokenCredential;
+    private ContainerRegistryTokenRequestContextImpl tokenRequestContext;
+    private final Predicate<AccessToken> shouldRefresh;
+    private final ClientLogger logger = new ClientLogger(AccessTokenCacheImpl.class);
+
+    /**
+     * Creates an instance of AccessTokenCache with default scheme "Bearer".
+     *
+     * @param tokenCredential the credential to be used to acquire token from.
+     */
+    AccessTokenCacheImpl(ContainerRegistryTokenCredentialImpl tokenCredential) {
+        Objects.requireNonNull(tokenCredential, "The token credential cannot be null");
+        this.wip = new AtomicReference<>();
+        this.tokenCredential = tokenCredential;
+        this.shouldRefresh = accessToken -> OffsetDateTime.now()
+            .isAfter(accessToken.getExpiresAt().minus(REFRESH_OFFSET));
+    }
+
+    /**
+     * Asynchronously get a token from either the cache or replenish the cache with a new token.
+     *
+     * @param tokenRequestContext The request context for token acquisition.
+     * @return The Publisher that emits an AccessToken
+     */
+    public Mono<AccessToken> getToken(ContainerRegistryTokenRequestContextImpl tokenRequestContext) {
+        return Mono.defer(retrieveToken(tokenRequestContext))
+            // Keep resubscribing as long as Mono.defer [token acquisition] emits empty().
+            .repeatWhenEmpty((Flux<Long> longFlux) -> longFlux.concatMap(ignored -> Flux.just(true)));
+    }
+
+    private Supplier<Mono<? extends AccessToken>> retrieveToken(ContainerRegistryTokenRequestContextImpl tokenRequestContext) {
+        return () -> {
+            try {
```

Comment:
This whole supplier should be wrapped with Mono.defer() to evaluate the need to refresh at subscription-time instead of at assembly-time.

Reply:
The caller actually does that - retrieveToken call just above this method. Since this is a private method chose not to do it.

---
In reply to: [597950212](https://github.com/Azure/azure-sdk-for-java/pull/19926#discussion_r597950212) [](ancestors = 597950212)